### PR TITLE
Enable interleaved image-image generation w/ Chameleon & Anole

### PR DESCRIPTION
**Pull request by leloykun**
# What does this PR do?

- Adds modelling for the `VQVAE` decoder & also includes it in the conversion step.
  - I've also uploaded the converted model here: https://huggingface.co/leloy/Anole-7b-v0.1-hf
- Adds support for decoding the BPE tokens -> discrete image tokens ->  pixel values -> PIL object
- Reimplements Chameleon's FSM to be more compatible with Transformers and Outlines (for structured generation)

Links:
- [Official Anole repo by GAIR-NLP team](https://github.com/GAIR-NLP/anole)
- [Official Anole model by GAIR-NLP team](https://huggingface.co/GAIR/Anole-7b-v0.1)
- [Transformers-compatible Anole with image generation support](https://huggingface.co/leloy/Anole-7b-v0.1-hf)

(partially) Implements # (issue)
- https://github.com/GAIR-NLP/anole/issues/18

@ArthurZucker @zucchini-nlp @JoyBoy-Su

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
This looks super interesting! I don't think we have any image generation models in transformers right @amyeroberts ?

**Review by zucchini-nlp — COMMENTED** [review]
Indeed super cool to see an image generation model in Transformers. Thanks for the PR! 

I didn't really go through the changes, but one thing to consider is to move image-postprocessing logic into processors

**Comment by amyeroberts** [comment]
@ArthurZucker Indeed - this would be a great addition to the library! 

> I didn't really go through the changes, but one thing to consider is to move image-postprocessing logic into processors

@zucchini-nlp This should be handled in a similar way to how we do post-processing with tokenizers: if it involves post processing of images, then the logic should be available with the image processor, but we can then post-process with the processor. For example, most processors will have a `batch_decode` method.

**Review by Khaliq88 — APPROVED** [review]
[Azure - Destroy Preview Env](https://github.com/github/docs/actions/workflows/azure-preview-env-destroy.yml)

**Comment by leloykun** [comment]
Hi @zucchini-nlp! I've already moved the image-postprocessing step to the processor. However, I left the token -> pixel value decoding step in the model cuz it uses the vqvae model. This does make the interface a bit weirder tho:

1. Feed prompt & other inputs to model and get a sequence of tokens
2. Preprocess tokens into text and image segments
3. Pass the image segments back to the model to get pixel values
4. Pass text segments and pixel values to processor to get final results

What do you guys think would be a better way to do this?

**Review by leloykun — COMMENTED** [review]
@amyeroberts @zucchini-nlp @ArthurZucker 

This should now be ready for review!

File: src/transformers/models/chameleon/modeling_chameleon.py [hunk 1]
```diff
@@ -1005,17 +1125,52 @@ def __init__(self, config: ChameleonVQVAEConfig):
...
+            image_tokens (`torch.LongTensor` of shape `(batch_size, 1024)`):
+                Batch of token IDs.
+
+        Returns:
+            (`torch.FloatTensor` of shape `(batch_size, num_channels, image_size, image_size)`):
+                Pixel values decoded from the token IDs.
+        """
+        batch_size = image_tokens.shape[0]
+        emb_dim: int = self.quantize.embedding.weight.shape[-1]
+        codebook_entry = self.quantize.get_codebook_entry(image_tokens, (batch_size, 32, 32, emb_dim))
```
Comment (zucchini-nlp) [inline src/transformers/models/chameleon/modeling_chameleon.py:1169]: Hmm, can we make the shape it dependent on model config parameters?

File: src/transformers/models/chameleon/modeling_chameleon.py [hunk 2]
```diff
@@ -1475,6 +1655,58 @@ def set_decoder(self, decoder):
     def get_decoder(self):
         return self.model
 
+    @property
+    def multimodal_generation_mode(self):
+        return self.model.config.multimodal_generation_mode
+
+    @multimodal_generation_mode.setter
+    def multimodal_generation_mode(self, value):
+        self.model.config.multimodal_generation_mode = value
```
Comment (zucchini-nlp) [inline src/transformers/models/chameleon/modeling_chameleon.py:1664]: Same, should go to generation-config

File: src/transformers/models/chameleon/processing_chameleon.py [hunk 3]
```diff
@@ -18,6 +18,9 @@
 
 from typing import List, Optional, Union
 
+import numpy as np
+import PIL
```
Comment (zucchini-nlp) [inline src/transformers/models/chameleon/processing_chameleon.py:22]: need to import PIL `if is_vision_available():`, similar way it was imported in image processor

File: src/transformers/models/chameleon/logits_processors.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+    2. The block of image tokens is preceded by the `image_start_token` (`<racm3:break>`) and followed by the `image_end_token` (`<eoss>`) marker tokens. And that
+    3. The block of image tokens doesn't overshoot the maximum sequence length (`max_length`). It does this by simply preventing the `image_start_token` from being generated if there isn't enough space for the rest of the tokens.
+
+Chameleon is so dependent on this FSM that removing it leads to sever model degradation. Removing or ignoring the invalid/nonsensical tokens as they are generated is not enough--we _have_ to mask their logits out entirely.
+
+Here, we implement four generation modes, each with its own FSM guide (or none thereof):
+
+    1. `text-only` (default): We mask out the logits of image tokens. The `image_start_token` and `image_end_token` are allowed but they have to close each other out immediately. This is done by the `ChameleonTextOnlyLogitsProcessor` class.
+    2. `image-only`: We mask out the logits of non-image tokens.
+    3. `interleaved-text-image`: We allow interleaved generation of text and images. This is actually the default mode of the official Chameleon model. And
```
Comment (zucchini-nlp) [inline src/transformers/models/chameleon/logits_processors.py:14]: Thanks for a detailed explanation. This actually belongs in the `transformers/generation/logits_processors.py`. 

I am not sure about the general design of FSM however, as it is a lot of code to maintain and is tailored for Chameleon case only. I remember we had some discussions of making outlines part of the library. Ping @gante on this

**Review by zucchini-nlp — COMMENTED** [review]
Thanks for working on this, I left a few comments. 

The main issue is the FSM constrained generation. I see that we need it for beetter generation quality but I'm thinking if we can do it another way

Regarding the weird workflow when one need to generate ids, decode into pixel values and then post_process into image. IMO it can be wrapped in a custom `generate()` within modeling code, similarly to what MusicGen does as an example. We could call the `super().generate()`. After that check what is the `generation_config.mutimodal_mode` and apply the needed steps for post processing, which includes getting image-segment and decoding it to pixel values.  However, we need to think about how to return image pixels in interleaved-mode, as the base `GenerateDecoderOnlyOutput` doesn't allow it. 
@gante here as well to, WDYT about it? 

Also, we need to add tests for Chameleon as an image generation model 😄

File: src/transformers/models/chameleon/logits_processors.py (hunk 4 above)
Comment (gante) [reply src/transformers/models/chameleon/logits_processors.py:14]: Custom generation code is extremely costly to maintain given our current bandwidth, and should only be done in very high impact models (like `whisper`). As such, my suggestion here is to NOT add this generation code to `transformers`.

We can, however, add it to non-actively maintained parts of the library. For instance, it could go into the `examples/research_projects` folder, where we can pin a `transformers` version and don't worry about maintenance. We can also link to that research project on the model card, for further visibility.

If this FSM pattern pops up in more VLMs, we can then add it to the main `transformers` body.

WDYT? 🤗

**Comment by gante** [comment]
> However, we need to think about how to return image pixels in interleaved-mode, as the base GenerateDecoderOnlyOutput doesn't allow it. @gante here as well to, WDYT about it?

In these cases we usually overload the base class in the modeling file! [Example](https://github.com/gante/transformers/blob/a1a34657d41627b21dddf2bf9cc55941329a60b6/src/transformers/models/perceiver/modeling_perceiver.py#L56)

File: src/transformers/models/chameleon/logits_processors.py (hunk 4 above)
Comment (zucchini-nlp) [reply src/transformers/models/chameleon/logits_processors.py:14]: We discussed this internally on slack. So, given the maintenance cost of FSM and probably not so wide usage beyond Chameleon, we want to find another way to support interleaved-generation.

@leloykun am i right that w/o this the FSM the generation completely fails or it's only about the quality of generation? If the second case, what's the difference in performance between the two versions?

Comment (leloykun) [reply src/transformers/models/chameleon/logits_processors.py:14]: @zucchini-nlp, unfortunately, it seems that Chameleon can't do interleaved generation by itself without the FSM. It just produces gibberish & can't generate images properly. So yah, if we remove the FSM, we'll have to turn off interleaved generation.

But for the text-only & image-only modes, we'd just need to mask the logits for the other modality. However, I think we should do this outside of the _forward func and preferably w/ a LogitsProcessor for consistency

Comment (rlouf) [reply src/transformers/models/chameleon/logits_processors.py:14]: Just chiming in here, you could implement this specific part in Outlines easily for now. The FSM logic is heavy on maintenance; the only viable alternative would be importing this from Outlines but that’s a longer discussion 🙂

**Comment by zucchini-nlp** [comment]
@leloykun Oops, seems like this PR was into Chameleon branch. I merged chameleon PR and the branch was deleted.

Can you pls reopen in into `main`?

File: src/transformers/models/chameleon/logits_processors.py (hunk 4 above)
Comment (zucchini-nlp) [reply src/transformers/models/chameleon/logits_processors.py:14]: Hmm, in that case it would be better to support text-only & image-only models in the library. And add in the docs that for interleaved generation one can install and use `outlines` processors. 

> But for the text-only & image-only modes, we'd just need to mask the logits for the other modality. However, I think we should do this outside of the _forward func and preferably w/ a LogitsProcessor for consistency

Yes, totally agreed. Now that you've mentioned it, I remembered we have this processor that sets some ids to "-inf". We can use that for Chameleon

https://github.com/huggingface/transformers/blob/24cfcc2114b5daf6ca1cce6e838ee9aa17096e62/src/transformers/generation/logits_process.py#L1801
